### PR TITLE
Lr/increment epoch

### PR DESCRIPTION
**Pull request by lukaszrzasik**
Part of #3730

### This PR: 

* Adds logic to update epoch if relevant when views are updated
* Adds view and epoch updating logic when Quorum Certificate is formed

### CODE REVIEW

File: crates/task-impls/src/quorum_proposal_recv/handlers.rs [hunk 1]
```diff
@@ -95,7 +97,14 @@ async fn validate_proposal_liveness<TYPES: NodeType, I: NodeImplementation<TYPES
...
 
-    if let Err(e) = update_view::<TYPES, I, V>(view_number, event_sender, task_state).await {
+    let epoch_number = TYPES::Epoch::new(epoch_from_block_number(
+        proposal.data.block_header.block_number(),
+        task_state.epoch_height,
+    ));
+
+    if let Err(e) =
+        update_view::<TYPES, I, V>(view_number, epoch_number, event_sender, task_state).await
+    {
```
Comment (bfish713) [inline crates/task-impls/src/quorum_proposal_recv/handlers.rs:107]: When you merge this with main this will just become a `ViewChange` with the new epoch number, and the changes in `update_view` will just have to be in `handle_view_change`

**Review by bfish713 — COMMENTED** [review]
This looks pretty straightforward, I'll take a look tomorrow once main is merged in

File: crates/task-impls/src/da.rs [hunk 2]
```diff
@@ -307,6 +307,18 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> DaTaskState<TYP
...
+                }
+
+                // If we are not the next leader (DA leader for this view) immediately exit
+                ensure!(
+                    self.da_membership
+                        .leader(self.cur_view + 1, self.cur_epoch)?
+                        == self.public_key
+                );
+
+                tracing::debug!("Polling for DA votes for view {}", *self.cur_view + 1);
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/da.rs]: I think my PR removed this code and that's why you got a merge conflict here.  I removed it because it was doing nothing

File: crates/task-impls/src/quorum_vote/mod.rs [hunk 3]
```diff
@@ -221,14 +217,42 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES> + 'static, V: Versions> Handl
...
+            *next_leaf_epoch
+        );
+        broadcast_event(
+            Arc::new(HotShotEvent::ViewChange(
+                self.view_number + 1,
+                next_leaf_epoch,
+            )),
+            &self.sender,
+        )
+        .await;
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/quorum_vote/mod.rs:246]: I think it should not matter much, we could even maybe just put the view change in the `QuorumVoteDependenciesValidated` handler

[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/quorum_vote/mod.rs:246]: Or even `QuorumVoteSend` handler.  But either way this should be ok

File: crates/task-impls/src/quorum_vote/mod.rs [hunk 4]
```diff
@@ -178,11 +179,6 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES> + 'static, V: Versions> Handl
             &self.sender,
         )
         .await;
-        broadcast_event(
-            Arc::new(HotShotEvent::ViewChange(self.view_number + 1)),
-            &self.sender,
-        )
-        .await;
```
Comment (bfish713) [inline crates/task-impls/src/quorum_vote/mod.rs:185]: This might be causing the regression

File: crates/task-impls/src/quorum_vote/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/quorum_vote/mod.rs:246]: Ok one problem is that some of the previous stuff is faillable, so we might not emit the event in some cases we did before, like we didn't see the last proposal and can't fetch it in time, does the epoch number calculation depend on the state update?

**Review by ss-es — COMMENTED** [review]
some random comments, looks fine on the whole

have some longer questions I will ask on zulip (about the `quorum_vote` and `consensus.rs` files)

File: crates/task-impls/src/quorum_vote/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/quorum_vote/mod.rs]: Yes we most switch views here, switching views is basically the same as starting the timeout for the next view.  If we don't do it here and the next proposal doesn't come in are behavior will not be correct

File: crates/types/src/consensus.rs [hunk 5]
```diff
@@ -882,100 +884,125 @@ impl<TYPES: NodeType> Consensus<TYPES> {
...
     }
 
-    /// Return true if the high QC takes part in forming an eQC, i.e.
-    /// it is one of the 3-chain certificates but not the eQC itself
-    pub fn is_high_qc_forming_eqc(&self) -> bool {
-        self.is_high_qc_for_last_block() && !self.is_high_qc_extended()
+    /// Returns an `Epoch` for the view following the given view.
+    /// There are three cases:
+    /// 1. The leaf from the given view is not for the last block.
+    ///    The epoch should be calculated based on the next block number.
```
Comment (bfish713) [inline crates/types/src/consensus.rs:974]: I don't think I understand this case, if the leaf is not the last block (say it's height 3 and epoch length is 100), then isn't this the trivial case and can just be handled without going into the validated state map?  We already got the height before calling this function right?  

Basically I expect the code on forming eqc to just do the decide rule, and if the block decided is the last of the epoch we can advance, all other cases we stay in the current epoch.  I can see how the logic kind of does that, but it's hard to follow when the case for "I'm not even trying to forma an eqc" and "I just formed an eqc" end up being handled by the same line of code and It's a little confusing that `is_leaf_forming_eqc` is false when we form the eqc?

File: crates/task-impls/src/quorum_proposal_recv/mod.rs [hunk 6]
```diff
@@ -169,6 +179,61 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions>
...
+                    "Sending ViewChange for view {} and epoch {}",
+                    *new_view,
+                    *next_view_epoch
+                );
+                broadcast_event(
+                    Arc::new(HotShotEvent::ViewChange(new_view, next_view_epoch)),
+                    &event_sender,
+                )
+                .await;
+            }
```
Comment (bfish713) [inline crates/task-impls/src/quorum_proposal_recv/mod.rs:236]: I thought we discussed only doing the epoch change when we get the proposal containing this eQC.  Idk how much it matters, and I think we'll discuss today the timeout path where this becomes relevant.  It's mostly a matter of what epoch do I want to timeout to and when.  I don't have a problem with this way, just noting.

**Review by bfish713 — APPROVED** [review]
